### PR TITLE
feat(1958): Upgrade to @hapi/hapi v19.x.x. BREAKING CHANGE: pulling in joi v17

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

Screwdriver hapi.js dependencies are outdated and not supported anymore.

## Objective

This PR majorly upgrades @hapi/joi version and fixes all schema related changes to be compatible with latest @hapi js dependencies.

## References

https://github.com/screwdriver-cd/screwdriver/issues/1958

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: api/auth.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-const Joi = require('joi');
+const Joi = require('@hapi/joi');
```
Comment (adong) [inline api/auth.js:3]: Just want to point out.
image


Ref: https://github.com/sideway/joi/issues/2411

Comment (jithine) [reply api/auth.js:3]: from what I read going to v19 is the first step, so this change will be required

Comment (jithine) [reply api/auth.js:3]: but you are right, we can omit `hapi@` module and just use `joi` with v17

File: api/commandValidator.js [hunk 2]
```diff
@@ -1,32 +1,34 @@
 'use strict';
 
-const Joi = require('joi');
+const Joi = require('@hapi/joi');
```
Comment (jithine) [inline api/commandValidator.js:3]: this can me reverted to `require('joi');` to avoid future changes.

Comment (pritamstyz4ever) [reply api/commandValidator.js:3]: I think we should wait until this is resolved.
https://github.com/sideway/joi/issues/2411

Comment (jithine) [reply api/commandValidator.js:3]: so what he's saying is  that ```The npm joi module is being resurrected and the current v17.1.1 has been published there (identical to the version available on @hapi/joi).```
current `@hapi/joi` & `joi` in npm are same

File: config/job.js [hunk 3]
```diff
@@ -3,25 +3,24 @@
...
-// ref. https://github.com/hapijs/joi/blob/v13.3.0/API.md#extendextension
+// ref. https://github.com/sideway/joi/blob/v17.1.1/API.md#extendextensions
 const sdJoi = Joi.extend(joi => ({
+    type: 'string',
     base: joi.string(),
-    name: 'string',
-    language: {
-        branchFilter: 'invalid trigger format'
+    messages: {
+        'string.branchFilter': '"{{#label}}" has invalid trigger format {{#q}}'
```
Comment (tkyi) [inline config/job.js:15]:
```suggestion
        'string.branchFilter': '{{#label}} has invalid trigger format {{#q}}'
```

File: core/scm.js [hunk 4]
```diff
@@ -164,14 +166,18 @@ const SCHEMA_HOOK = Joi.object().keys({
 
     ref: Joi.string()
         .when('action',
-            { is: ['release', 'tag'], then: Joi.required(), otherwise: Joi.allow('').optional() })
+            {
+                is: Joi.valid('release', 'tag'),
+                then: Joi.required(),
+                otherwise: Joi.optional()
```
Comment (tkyi) [inline core/scm.js:172]: missing `allow('')` for otherwise case?

File: models/banner.js [hunk 5]
```diff
@@ -33,12 +33,7 @@ const MODEL = {
         .description('Username of user creating the banner')
         .example('batman123'),
 
-    type: Joi
-        .string().valid([
-            'info',
-            'warn'
-        ])
-        .max(32)
+    type: Joi.string().valid('info', 'warn')
```
Comment (tkyi) [inline models/banner.js:36]: might still need max length for https://github.com/screwdriver-cd/screwdriver/issues/325?
actually nvm looks like this PR should check: https://github.com/screwdriver-cd/data-schema/pull/280/files
```suggestion
    type: Joi.string().valid('info', 'warn').max(32)
```

Comment (pritamstyz4ever) [reply models/banner.js:36]: @tkyi The .max() function is not allowed for .valid(). It has two values that should validate only info and warn.

File: config/template.js [hunk 6]
```diff
@@ -56,27 +57,21 @@ const TEMPLATE_MAINTAINER = Joi
...
-        config: Job.job,
+        name: TEMPLATE_NAME.required(),
+        version: TEMPLATE_VERSION.required(),
+        description: TEMPLATE_DESCRIPTION.required(),
+        maintainer: TEMPLATE_MAINTAINER.required(),
+        config: Job.job.required(),
         images: TEMPLATE_IMAGES
-    })
-    .requiredKeys('name', 'version', 'description', 'maintainer',
-        'config', 'config.image', 'config.steps');
```
Comment (tkyi) [inline config/template.js:79]: is there any way to account for the requires `config.image` and `config.steps`?

Comment (pritamstyz4ever) [reply config/template.js:79]: Added the check for `config.image` and `config.steps`. 
Although, there is hardly any documentation for requiredKeys alternative in hapi 17, we may need to update some base schemas in some cases.

**Pull request merged**